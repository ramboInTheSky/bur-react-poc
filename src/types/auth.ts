import { User } from './user';

export interface Auth {
  username: string;
  password: string;
  profile?: User;
}
