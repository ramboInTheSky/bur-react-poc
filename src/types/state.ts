import { User } from './user';
import { Forms } from './forms';

export interface State {
    user: User | undefined;
    forms: Forms;
}