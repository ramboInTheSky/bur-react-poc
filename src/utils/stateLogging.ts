export default (state: any, action: any) => {
    const env = process.env.NODE_ENV;
    if (env !== 'production' && env !== 'test') {
        console.groupCollapsed('%c Application state change', 'color: white; background: black; padding: 2px');
        console.log(`%cAction: ${action.type}`, 'color: purple');
        console.log('%cState: ', 'color: blue', state);
        console.groupEnd();
    }
}