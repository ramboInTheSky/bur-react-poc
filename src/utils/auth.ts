import { Auth } from '../types/auth';

const {
  getAccountInfo,
  initRegistration,
  login: sdkLogin,
  logout: sdkLogout,
  register: sdkRegister,
} = window.gigya.accounts;

export const login = ({ username, password }: Auth) => {
  return new Promise((resolve, reject) => {
    sdkLogin({
      loginID: username,
      password,
      callback: (response: any) => {
        if (response.errorCode === 0) {
          resolve(response.user);
        } else {
          reject('Something went wrong');
        }
      },
    });
  });
};

export const logout = () => {
  return new Promise((resolve, reject) => {
    sdkLogout({
      callback: (response: any) => {
        if (response.errorCode === 0) {
          resolve();
        } else {
          reject('Something went wrong');
        }
      },
    });
  });
};

export const register = ({ username, password, profile }: Auth) => {
  return new Promise((resolve, reject) => {
    initRegistration({
      callback: (response: any) => {
        if (response.errorCode === 0) {
          sdkRegister({
            email: username,
            password,
            profile,
            regToken: response.regToken,
            finalizeRegistration: true,
            format: 'jsonp',
            callback: (response: any) => {
              if (response.errorCode === 0) {
                resolve(response.user);
              } else {
                reject('Something went wrong');
              }
            },
          });
        } else {
          reject('Something went wrong');
        }
      },
    });
  });
};

export const getAccount = () => {
  return new Promise((resolve) => {
    getAccountInfo({
      callback: (response: any) => {
        if (response.errorCode === 0) {
          resolve(response.profile);
        } else {
          resolve(undefined);
        }
      },
    });
  });
}
