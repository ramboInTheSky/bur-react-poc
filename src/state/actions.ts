import { login as authLogin, register as authRegister } from '../utils/auth';
import { User } from '../types/user';

export default (dispatch: any) => ({
  login: (username: string, password: string, redirectUrl?: Location) =>
    authLogin({
      username,
      password,
    }).then((user: any) => {
      dispatch({ type: 'updateUser', payload: user });
      if (redirectUrl && user) {
        window.location = redirectUrl;
      }
    }),

  logout: () => dispatch({ type: 'updateUser', payload: undefined }),

  register: (username: string, password: string, profile: User, redirectUrl?: Location) =>
    authRegister({
      username,
      password,
      profile,
    }).then((user: any) => {
      dispatch({ type: 'updateUser', payload: user });
      if (redirectUrl && user) {
        window.location = redirectUrl;
      }
    }),
});
