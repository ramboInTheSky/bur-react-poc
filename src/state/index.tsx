import React, { createContext, useContext, useReducer } from 'react';
import { State } from '../types/state';

export const initialState: State = {
  user: undefined,
  forms: {
    username: '',
    firstname: '',
    lastname: '',
    password: '',
    consent: false,
  },
};

export const StateContext = createContext({});

export const StateProvider = (props: any) => {
  const { reducer, children, state = initialState} = props;
  return (
    <StateContext.Provider value={useReducer(reducer, state)}>
      {children}
    </StateContext.Provider>
  );
};

export const useStateValue = () => useContext(StateContext);
