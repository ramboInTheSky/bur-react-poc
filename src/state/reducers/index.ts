import logging from '../../utils/stateLogging';
import { State } from '../../types/state';

import userReducer from './user';
import formsReducer from './forms';

export default (state: State, action: any) => {
  logging(state, action);
  return {
    user: userReducer(state.user, action),
    forms: formsReducer(state.forms, action),
  };
};
