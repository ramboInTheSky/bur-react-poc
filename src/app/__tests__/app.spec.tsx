import React from 'react';
import { shallow, mount } from 'enzyme';

import { StateProvider } from '../../state';
import reducer from '../../state/reducers';

import App from '../app';

import * as auth from '../../utils/auth';
jest.mock('../../utils/auth', () => ({
  getAccount: jest.fn(() => Promise.resolve({ email: 'test@test.com' })),
}));

// TODO
// Get rid of act() error in terminal
// Test setting global state

describe(App, () => {
  let comp;

  beforeEach(() => {
    comp = (
      <StateProvider reducer={reducer}>
        <App />
      </StateProvider>
    );
  });
  it('should render correctly', () => {
    expect(shallow(comp)).toMatchSnapshot();
  });

  it('should attempt to get account details on mount', () => {
    mount(comp);
    expect(auth.getAccount).toHaveBeenCalled();
  });
});
