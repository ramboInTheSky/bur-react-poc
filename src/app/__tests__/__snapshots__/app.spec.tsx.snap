// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<ContextProvider
  value={
    Array [
      Object {
        "forms": Object {
          "consent": false,
          "firstname": "",
          "lastname": "",
          "password": "",
          "username": "",
        },
        "user": undefined,
      },
      [Function],
    ]
  }
>
  <App />
</ContextProvider>
`;
